Redo the Pass Manager

CodaFi opened the pull request:
This thing was never correct.  We mixed up Function passes and Module passes, and if you were unlucky enough to schedule both into the function pass manager then LLVM would probably crash.

Provide the interface we were shooting for here anyways in the form of a Swift-like pass pipeliner.  You can freely mix module and function-level passes in the pipeliner, and we will schedule them the way LLVM would anyways (group the function passes, then run the module passes).  There's bindings for LTO passes that never made it into a header that we could also wrap with this infrastructure.

CodaFi committed:
Redo the pass manager

matthewseaman commented on Sources/LLVM/PassPipeliner.swift:
```diff
@@ -0,0 +1,299 @@
+         .globalDCE,
+         .deadArgElimination,
+         .ipConstantPropagation,
+         .ipscc:
```
I believe at least .functionAttrs should be here too.

CodaFi committed:
Allow executing with a pipeline mask

matthewseaman commented on Sources/LLVM/PassManager.swift:
```diff
@@ -106,8 +106,6 @@ public enum FunctionPass {
   /// Replace aggregates or pieces of aggregates with scalar SSA values.
   case scalarReplAggregatesSSA
-  /// Tries to inline the fast path of library calls such as sqrt.
-  case simplifyLibCalls
```
Why is this removed?

CodaFi replied:
The corresponding add pass function is a no-op as the pass has been removed.

matthewseaman reviewed (commented):
Looking great!

CodaFi committed:
Add the "internalize" and "add discriminators" passes

CodaFi committed:
Deprecate simplifyLibCalls and unify SROA

All the SROA variants in the C API no longer make a difference.

CodaFi committed:
Redo the pass configuration infrastructure

This finally allows us to wrap passes that require additional parameters or configuration calls.

In order to do this for internalize in particular, we have to do some heroic lifetime juggling.

CodaFi committed:
Add tests for internalize

CodaFi commented:
⛵️
